File: python3/LeetCode/xorQueries.py
from typing import List


class Solution:
    def xorQueries(self, arr:List[int], queries:List[List[int]]) -> List[int]:
        n = len(arr)
        dp = [0] * n
        dp[0] = arr[0]
        for i in range(1, n):
            dp[i] = dp[i-1] ^ arr[i]

        res = [0] * len(queries)
        for i in range (0, len(queries)):
            if queries[i][0] == 0:
                res[i] = dp[queries[i][1]]
            else:
                res[i] = dp[queries[i][0]-1] ^ dp[queries[i][1]]

        return res




if __name__ == '__main__':
    arr = [1,3,4,8]
    queries =  [[0,1],[1,2],[0,3],[3,3]]
    print(Solution().xorQueries(arr, queries))

File: go/LeetCode/xorQueries.go
package main

import "fmt"

func main() {
	arr := []int{4,8,2,10}
	queries := [][]int{{2,3},{1,3},{0,0},{0,3}}
	fmt.Print(xorQueries(arr, queries))
}

func xorQueries(arr []int, queries [][]int) []int {
	n:= len(arr)
	dp := make([]int, n)
	dp[0] = arr[0]
	for i:=1;i<n;i++ {
		dp[i] = dp[i-1] ^ arr[i]
	}

	res := make([]int, len(queries))
	for i,v := range queries{
		if v[0] == 0 {
			res[i] = dp[v[1]]
		}else {
			res[i] = dp[v[0]-1] ^ dp[v[1]]
		}
	}
	return res
}
